File: Data Structures & Algorithms/binary-search/submission-1.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int index = -1;
        int left = 0, right = nums.size();

        while (left < right) {
            int mid = (left + right) / 2;
            if (nums[mid] == target) return mid;
            else if (nums[mid] < target) left = mid + 1;
            else right = mid;
        }

        return index;
    }
};

File: Data Structures & Algorithms/path-sum/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        if not root:
            return False
        
        def rec(node, target):
            assert node, "node must not be None"
            target -= node.val
            if not node.left and not node.right:
                return target == 0
            left = rec(node.left, target) if node.left else False
            right = rec(node.right, target) if node.right else False
            return left or right
        return rec(root, targetSum)